[Merged by Bors] - feat(AlgebraicTopology/SimplicialSet): StrictSegal simplicial sets are 2-coskeletal

emilyriehl opened the pull request:
We prove that if `X` is a simplicial set that is `StrictSegal` then `X` is 2-coskeletal, meaning `(coskAdj 2).unit.app X` defines an isomorphism `StrictSegal.cosk2Iso : X ≅ (Truncated.cosk 2).obj ((truncation 2).obj X)`. 

This is proven by demonstrating that `(rightExtensionInclusion X n).IsPointwiseRightKanExtension` holds. 

Co-Authored-By: [Mario Carneiro](https://github.com/digama0) and [Joël Riou](https://github.com/joelriou) 

---
- [ ] depends on: #19492 

This replaces #16782, half of which has already been merged.

Note this file does not currently contain a definition of `n-coskeletal` because it wasn't clear to me which form of that definition would be most useful. My guess is that it should be something like `IsIso ((coskAdj n).unit.app X)`. Please let me know if this is something you'd like me to add or if you have other suggestions.

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,441 @@
+
+/-- A strict Segal simplicial set is 2-coskeletal. -/
+noncomputable def IsPointwiseRightKanExtensionAt (n : ℕ) :
+    (rightExtensionInclusion X 2).IsPointwiseRightKanExtensionAt ⟨[n]⟩ := by
+  show IsLimit _
```
When defining data, it is not recommended to apply too many tactics, especially `simp`. The structure of the code should be like this:
```suggestion
    (rightExtensionInclusion X 2).IsPointwiseRightKanExtensionAt ⟨[n]⟩ where
  lift s x := ran.lift (X := X) s x
  fac := sorry
  uniq := sorry
```

emilyriehl replied:
Done (with your branch).

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,441 @@
+variable (X : SSet.{u}) [StrictSegal X]
+
+/-- A strict Segal simplicial set is 2-coskeletal. -/
+noncomputable def IsPointwiseRightKanExtensionAt (n : ℕ) :
```
```suggestion
noncomputable def isPointwiseRightKanExtensionAt (n : ℕ) :
```

emilyriehl replied:
Done (with your branch).

joelriou commented:
I have tried to improve the code in the branch https://github.com/leanprover-community/mathlib4/tree/coskeletal_jriou
see https://github.com/leanprover-community/mathlib4/blob/coskeletal_jriou/Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean (which is still a kind of draft, and there is still a not so hard sorry).
Mostly, we may avoid introducing too many auxiliary definitions. Also, when checking equalities between `2`-simplices of strict Segal simplicial sets, it suffices to check two equalities between arrows (as the equalities between vertices can be deduces from these): see the additional ext lemmas at https://github.com/leanprover-community/mathlib4/blob/coskeletal_jriou/Mathlib/AlgebraicTopology/SimplicialSet/Path.lean

joelriou commented:
> see the additional ext lemmas at https://github.com/leanprover-community/mathlib4/blob/coskeletal_jriou/Mathlib/AlgebraicTopology/SimplicialSet/Path.lean
Actually, these should be generalized as an `ext` lemma for paths of length `n + 1`.

joelriou commented:
In general, could you tag your PR with `t-category-theory` rather than `t-topology`? We do not have `t-algebraic-topology` (nor `t-homology`), but these constructions on simplicial sets, etc, fit much more the `t-category-theory` tag as compared to `t-topology` which is very much related to topological spaces. (Also, `t-topology` is mostly under my radar, but I follow quite closely the contributions in `t-category-theory`.)

emilyriehl commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,441 @@
+/-- The identity natural transformation exhibits a simplicial set as a right extension of its
+restriction along `(Truncated.inclusion (n := n)).op`.-/
+@[simps!]
+def rightExtensionInclusion (X : SSet.{u}) (n : ℕ) :
```
Done.

gio256 commented:
@emilyriehl I apologize, I should have warned you about this, but #19349 introduced the lemma suggested by @joelriou to the `master` branch. This is likely the source of your merge conflict now.

I'm happy to help resolve this, although I'm not positive what the best route is. Normally, I would open a PR against your fork resolving the conflict. In this case, I suspect a PR opened against this branch might require a maintainer to merge. I *think* I might be able to push directly to your branch, but that would obviously require some coordination to make sure I don't push anything that conflicts with your work.

emilyriehl commented:
Thanks @gio256. The conflict was silly (just two new lemmas in the same place) but your `Path.ext` seemed better/more general than the versions `Path.ext_1` and `Path.ext_2` that @joelriou had defined in a branch of this PR. So I cut those and fixed everything (I think). But if I'm still stuck, I may take you up on your offer of help. Thanks.

emilyriehl commented:
@joelriou I've
- merged your branch
- filled in that sorry (with a proof I'm sure needs golfing...)
- bumped mathlib (and replaced your two Path.ext lemmas with `Path.ext'`)
- implemented all the suggestions above with one exception I'll now describe.

I'll now open up a new PR which defines IsCoskeletal at the level of simplicial objects. I'm hoping that will simplify and perhaps speed up the review.

emilyriehl commented:
Update: see #19492.

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,270 @@
+    · exact congr_fun (hm (StructuredArrow.mk (Y := op [0]₂) ([0].const [n] i).op)) x
+    · exact congr_fun (hm (.mk (Y := op [1]₂) (.op (mkOfLe _ _ (Fin.castSucc_le_succ i))))) x
+
+/-- Since `rightExtensionInclusion₂IsPointwiseRightKanExtensionAt X n` proves that the appropriate
```
```suggestion
/-- Since `StrictSegal.isPointwiseRightKanExtensionAt` proves that the appropriate
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,270 @@
+/-- Since `rightExtensionInclusion₂IsPointwiseRightKanExtensionAt X n` proves that the appropriate
+cones are limit cones, `rightExtensionInclusion X 2` is a pointwise right Kan extension.-/
+noncomputable def isPointwiseRightKanExtension :
+    RightExtension.IsPointwiseRightKanExtension (rightExtensionInclusion X 2) :=
```
We can use dot notation:
```suggestion
    (rightExtensionInclusion X 2).IsPointwiseRightKanExtension :=
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,270 @@
+  fun Δ => isPointwiseRightKanExtensionAt X Δ.unop.len
+
+theorem isRightKanExtension :
+    (rightExtensionInclusion X 2).left.IsRightKanExtension (rightExtensionInclusion X 2).hom :=
```
```suggestion
    X.IsRightKanExtension (𝟙 ((inclusion 2).op ⋙ X)) :=
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,270 @@
+instance isCoskeletal : SimplicialObject.IsCoskeletal X 2 where
+  isRightKanExtension := isRightKanExtension X
+
+/-- The canonical isomorphism `X ≅ (cosk 2).obj X` defined when `X` is `StrictSegal`. -/
+noncomputable def isoCosk2 : X ≅ (cosk 2).obj X := SimplicialObject.isoCoskOfIsCoskeletal X 2
+
```
This can be removed: because of the typeclass system, when `X` is strict Segal, it is 2-coskeletal due to the instance above, and then the isomorphism is just `X.isoCoskOfIsCoskeletal 2`.
```suggestion
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,270 @@
+
+open SSet
+
+instance isCoskeletal (C : Type u) [Category.{v} C] :
+    SimplicialObject.IsCoskeletal (nerve C) 2 := by infer_instance
```
When something is already an instance (here because we already now `nerve C` is strict Segal), it is not necessary to declare it, but in order to point out a fact to the reader, it is advisable to use `example`:
```suggestion
example (C : Type u) [Category.{v} C] :
    SimplicialObject.IsCoskeletal (nerve C) 2 := by infer_instance
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,270 @@
+/-- The natural isomorphism between `nerveFunctor` and `nerveFunctor ⋙ Truncated.cosk 2` whose
+components `nerve C ≅  (Truncated.cosk 2).obj (nerve₂ C)` shows that nerves of categories are
+2-coskeletal.-/
+noncomputable def cosk2Iso : nerveFunctor.{u, u} ≅ nerveFunctor₂.{u, u} ⋙ Truncated.cosk 2 := by
+  refine NatIso.ofComponents ?_ ?_
+  · intro C
+    dsimp [nerveFunctor, nerveFunctor₂]
+    exact (StrictSegal.isoCosk2 (nerve C))
+  · simp only [nerveFunctor_obj, comp_obj, id_eq, Functor.comp_map]
+    convert cosk2NatTrans.naturality
```
```suggestion
/-- The natural isomorphism between `nerveFunctor` and `nerveFunctor ⋙ Truncated.cosk 2` whose
components `nerve C ≅ (Truncated.cosk 2).obj (nerve₂ C)` shows that nerves of categories are
2-coskeletal.-/
noncomputable def cosk₂Iso : nerveFunctor.{u, v} ≅ nerveFunctor₂.{v, u} ⋙ Truncated.cosk 2 :=
  NatIso.ofComponents (fun C ↦ (nerve C).isoCoskOfIsCoskeletal 2)
    (fun _ ↦ (coskAdj 2).unit.naturality _)
```

Then, you may remove the definition `cosk2NatTrans`, because it is always an isomorphism.
You may also fix the definition I had made of `nerveFunctor` by making the universe explicit:
```lean
def nerveFunctor : Cat.{v, u} ⥤ SSet where
```
Then, this will fix the order of the universes, and in this file `Coskeletal.lean`, the definition would become:
```lean
noncomputable def cosk₂Iso : nerveFunctor.{v, u} ≅ nerveFunctor₂.{v, u} ⋙ Truncated.cosk 2 :=
```
(As this may be confusing, please let me know if this does not fully make sense to you.)

emilyriehl replied:
Done.

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,270 @@
+/-!
+# Coskeletal simplicial sets
+
+In this file, we prove that if `X` is `StrictSegal` then `X` is 2-coskeletal defining
+`X ≅ (cosk 2).obj X` In particular, nerves of categories are 2-coskeletal.
```
```suggestion
In this file, we prove that if `X` is `StrictSegal` then `X` is 2-coskeletal,
i.e. `X ≅ (cosk 2).obj X`. In particular, nerves of categories are 2-coskeletal.
```

joelriou commented on Mathlib/AlgebraicTopology/SimplexCategory.lean:
```diff
@@ -231,6 +231,10 @@ def mkOfLe {n} (i j : Fin (n+1)) (h : i ≤ j) : ([1] : SimplexCategory) ⟶ [n]
       | 0, 1, _ => h
   }
 
+@[simp]
+lemma mkOfLe_rfl_const {n} (j : Fin (n + 1)) :
+    mkOfLe j j (by omega) = [1].const [0] 0 ≫ [0].const [n] j := Hom.ext_one_left _ _
```
I think the RHS would be better in this version:
```suggestion
@[simp]
lemma mkOfLe_refl {n} (j : Fin (n + 1)) :
    mkOfLe j j (by omega) = [1].const [n] j := Hom.ext_one_left _ _
```
(This lemma does not seem to be used in the main result here.)

emilyriehl replied:
I've changed the lemma, as suggested, but the old version is used in the proof of the base case of fac_aux2 (see line 107) and I wasn't able to use the simpler version. Line 121 of the proof is 

```
refine congrArg (X.map ([1].const [0] 0).op) ?_
```
which uses the factorization of the constant map.

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Nerve.lean:
```diff
@@ -37,7 +37,7 @@ instance {C : Type*} [Category C] {Δ : SimplexCategoryᵒᵖ} : Category ((nerv
 /-- The nerve of a category, as a functor `Cat ⥤ SSet` -/
 @[simps]
-def nerveFunctor : Cat ⥤ SSet where
+def nerveFunctor : Cat.{v,u} ⥤ SSet where
```
```suggestion
def nerveFunctor : Cat.{v, u} ⥤ SSet where
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,260 @@
+
+lemma fac_aux₃ {n : ℕ}
+    (s : Cone (proj (op [n]) (Truncated.inclusion 2).op ⋙ (Truncated.inclusion 2).op ⋙ X))
+    (x : s.pt)
+    (φ : [1] ⟶ [n]) :
```
```suggestion
    (x : s.pt) (φ : [1] ⟶ [n]) :
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,260 @@
+        Int.reduceAdd, Fin.eta, comp_obj, StructuredArrow.proj_obj, op_obj, const_obj_obj,
+        const_obj_map, types_comp_apply, types_id_apply, Functor.comp_map, StructuredArrow.proj_map,
+        op_map] at nat
+      rw [nat]
+      rw [op_comp, Functor.map_comp]
```
```suggestion
      rw [nat, op_comp, Functor.map_comp]
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,260 @@
+
+lemma fac_aux₁ {n : ℕ}
+    (s : Cone (proj (op [n]) (Truncated.inclusion 2).op ⋙ (Truncated.inclusion 2).op ⋙ X))
+    (x : s.pt)
+    (i : ℕ) (hi : i < n) :
```
```suggestion
    (x : s.pt) (i : ℕ) (hi : i < n) :
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,260 @@
+recorded by the composite functor `nerveFunctor₂`.-/
+def nerveFunctor₂ : Cat.{v, u} ⥤ SSet.Truncated 2 := nerveFunctor ⋙ truncation 2
+
+/-- The natural isomorphism between `nerveFunctor` and `nerveFunctor ⋙ Truncated.cosk 2` whose
```
```suggestion
/-- The natural isomorphism between `nerveFunctor` and `nerveFunctor₂ ⋙ Truncated.cosk 2` whose
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,260 @@
+      let α : strArrowMk₂ f hi ⟶ strArrowMk₂ ([0].const [n] (f.toOrderHom k)) (by omega) :=
+        StructuredArrow.homMk (([0].const _ (by exact k)).op) (by simp; rfl)
+      exact congr_fun (s.w α).symm x
+    · dsimp only [spineEquiv, Equiv.coe_fn_mk]
+      dsimp
```
```suggestion
    · dsimp only [spineEquiv, Equiv.coe_fn_mk, spine_arrow]
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,256 @@
+
+lemma fac_aux₂ {n : ℕ}
+    (s : Cone (proj (op [n]) (Truncated.inclusion 2).op ⋙ (Truncated.inclusion 2).op ⋙ X))
+    (x : s.pt)
+    (i j : ℕ) (hij : i ≤ j) (hj : j ≤ n) :
```
```suggestion
    (x : s.pt) (i j : ℕ) (hij : i ≤ j) (hj : j ≤ n) :
```

joelriou commented:
maintainer merge

mattrobball commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,255 @@
+
+set_option quotPrecheck false
+local macro:max (priority := high) "[" n:term "]₂" : term =>
+  `((⟨SimplexCategory.mk $n, by decide⟩ : SimplexCategory.Truncated 2))
```
```suggestion
  `((⟨SimplexCategory.mk $n, by dsimp; omega⟩ : SimplexCategory.Truncated 2))
```
seems sufficient for this file. Is the `set_option` still needed?

mattrobball commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,255 @@
+
+open SSet
+
+example (C : Type u) [Category.{v} C] :
```
I don't understand the purpose of this example. If it is documentation, then no one will see it outside of a doc comment. If it is to make sure instance synthesis succeeds, then it should be a test importing more than this single file.

emilyriehl replied:
Hi @mattrobball. I don't understand what you mean by "...then it should be a test importing more than this single file." Could you explain?

That instance gets used implicitly in the final term `CategoryTheory.Nerve.cosk₂Iso` in the file. But as you guessed, instance synthesis succeeds. That term is still definable if this example is deleted.

On the other hand, to a human reader, this example and the consequent isomorphism is the whole point of this file (the main theorem). If it's consistent with style guidelines, I'd like to keep this as an explicit term so that when someone searches for this result later they'll be able to find it, without having to figure out that this is a special case of the more abstract theorem proven here (which is less "well-known"). I'd added an explicit mention to this fact in the comments at the top of the file.

Does this work for you, or would you like me to make further changes?

jcommelin replied:
I think it's fine like this.

mattrobball replied:
I'm not a big fan of `example`s in the code base because they tend to add to the imports and overall build time. They also confound environmental linters since nothing added to the environment.  

If keeping it as active code for documentation is important, I think declaring it as instance would cause less problems. But including the example as a code block in the comment itself would probably reach more eyes than doing this. 

Sometimes these examples are used to assure that instance synthesis doesn't fail under future chances. I don't think that is the intention here but I could be wrong. If it was, then one would want to assure that instance synthesis keeps working thoughout mathlib, ie nothing downstream breaks it. Having the example here does not assure this.

jcommelin replied:
Those are actually good points. Maybe we should codify this more in the contributor guidelines?

mattrobball replied:
One could also make it a `def` to avoid adding an instance but also making it visible to any environmental searching.  

> Those are actually good points. Maybe we should codify this more in the contributor guidelines?

I agree.

jcommelin replied:
Otoh, in this specific case we aren't adding extra imports, right?
I think there is value to having this as active code, even in this file.

It can also be added in the docstring, but since it is already mentioned in the module docstring I think we're good.

Let's not block this PR on this issue. We can always change/improve later.

bors merge

mattrobball replied:
We should make it an honest declaration then. It would come at no cost and some benefit.

mattrobball commented on Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean:
```diff
@@ -0,0 +1,255 @@
+  lift s x := lift (X := X) s x
+  fac s j := by
+    ext x
+    obtain ⟨⟨i, hi⟩, ⟨f⟩, rfl⟩ := j.mk_surjective
+    obtain ⟨i, rfl⟩ : ∃ j, SimplexCategory.mk j = i := ⟨_, i.mk_len⟩
+    change [i] ⟶ [n] at f
```
```suggestion
    obtain ⟨⟨i, hi⟩, ⟨f :  _ ⟶ _⟩, rfl⟩ := j.mk_surjective
    obtain ⟨i, rfl⟩ : ∃ j, SimplexCategory.mk j = i := ⟨_, i.mk_len⟩
```

mattrobball reviewed (commented):
Awesome seeing this stuff coming together 🎉

jcommelin reviewed (commented):
Thanks :tada:

bors merge
